File: packages/honker-bun/test/typed_jobs.test.ts
import { test, expect, describe } from "bun:test";
import { existsSync, mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join, resolve } from "node:path";

import { open, type Database } from "../src/index.ts";

const REPO_ROOT = resolve(import.meta.dir, "..", "..", "..");
const EXT_CANDIDATES = [
  "target/debug/libhonker_ext.dylib",
  "target/debug/libhonker_ext.so",
  "target/debug/libhonker_extension.dylib",
  "target/debug/libhonker_extension.so",
  "target/release/libhonker_ext.dylib",
  "target/release/libhonker_ext.so",
  "target/release/libhonker_extension.dylib",
  "target/release/libhonker_extension.so",
];

function findExtension(): string | null {
  const fromEnv = process.env.HONKER_EXT_PATH;
  if (fromEnv && existsSync(fromEnv)) return fromEnv;
  for (const rel of EXT_CANDIDATES) {
    const p = join(REPO_ROOT, rel);
    if (existsSync(p)) return p;
  }
  return null;
}

const extPath = findExtension();
if (!extPath && process.env.CI) {
  throw new Error("HONKER_EXT_PATH not found in CI; Bun tests must run for real");
}
const maybe = extPath ? describe : describe.skip;

interface EmailPayload {
  recipient: string;
  template: "welcome" | "receipt";
}

/** Read the clock the core stamps rows with, not the JS clock. */
function now(db: Database): number {
  return db.raw.query<{ v: number }, []>("SELECT unixepoch() AS v").get()!.v;
}

function withDb(fn: (db: Database) => Promise<void> | void): () => Promise<void> {
  return async () => {
    const dir = mkdtempSync(join(tmpdir(), "honker-bun-typed-jobs-"));
    const dbPath = join(dir, "t.db");
    const db = open(dbPath, extPath!);
    try {
      await fn(db);
    } finally {
      db.close();
      rmSync(dir, { recursive: true, force: true });
    }
  };
}

maybe("honker-bun job details", () => {
  test(
    "a claimed job carries every field the core returns",
    withDb((db) => {
      const queue = db.queue<EmailPayload>("emails", {
        visibilityTimeoutS: 120,
        maxAttempts: 5,
      });
      const runAt = now(db) - 5;

      const enqueuedBefore = now(db);
      const id = queue.enqueue(
        { recipient: "alice@example.com", template: "welcome" },
        { runAt, priority: 7, expires: 600 },
      );
      const enqueuedAfter = now(db);

      const claimedBefore = now(db);
      const job = queue.claimOne("worker-1");
      const claimedAfter = now(db);

      expect(job).not.toBeNull();
      expect(job!.id).toBe(id);
      expect(job!.queue).toBe("emails");
      expect(job!.payload).toEqual({
        recipient: "alice@example.com",
        template: "welcome",
      });
      expect(job!.state).toBe("processing");
      expect(job!.priority).toBe(7);
      expect(job!.runAt).toBe(runAt);
      expect(job!.workerId).toBe("worker-1");
      expect(job!.claimExpiresAt).toBeGreaterThanOrEqual(claimedBefore + 120);
      expect(job!.claimExpiresAt).toBeLessThanOrEqual(claimedAfter + 120);
      expect(job!.attempts).toBe(1);
      expect(job!.maxAttempts).toBe(5);
      expect(job!.createdAt).toBeGreaterThanOrEqual(enqueuedBefore);
      expect(job!.createdAt).toBeLessThanOrEqual(enqueuedAfter);
      expect(job!.expiresAt).toBeGreaterThanOrEqual(enqueuedBefore + 600);
      expect(job!.expiresAt).toBeLessThanOrEqual(enqueuedAfter + 600);
    }),
  );

  test(
    "a pending snapshot carries every field the core returns",
    withDb((db) => {
      const queue = db.queue<EmailPayload>("emails", { maxAttempts: 2 });
      const runAt = now(db) + 3600;

      const enqueuedBefore = now(db);
      const id = queue.enqueue(
        { recipient: "bob@example.com", template: "receipt" },
        { runAt, priority: 4, expires: 900 },
      );
      const enqueuedAfter = now(db);

      const snapshot = queue.getJob(id);
      expect(snapshot).not.toBeNull();
      expect(snapshot!.id).toBe(id);
      expect(snapshot!.queue).toBe("emails");
      expect(snapshot!.payload).toEqual({
        recipient: "bob@example.com",
        template: "receipt",
      });
      expect(snapshot!.state).toBe("pending");
      expect(snapshot!.priority).toBe(4);
      expect(snapshot!.runAt).toBe(runAt);
      expect(snapshot!.workerId).toBeNull();
      expect(snapshot!.claimExpiresAt).toBeNull();
      expect(snapshot!.attempts).toBe(0);
      expect(snapshot!.maxAttempts).toBe(2);
      expect(snapshot!.createdAt).toBeGreaterThanOrEqual(enqueuedBefore);
      expect(snapshot!.createdAt).toBeLessThanOrEqual(enqueuedAfter);
      expect(snapshot!.expiresAt).toBeGreaterThanOrEqual(enqueuedBefore + 900);
      expect(snapshot!.expiresAt).toBeLessThanOrEqual(enqueuedAfter + 900);
    }),
  );

  test("a reader sees the processing snapshot, then nothing after ack", async () => {
    const dir = mkdtempSync(join(tmpdir(), "honker-bun-typed-jobs-"));
    const dbPath = join(dir, "t.db");
    const worker = open(dbPath, extPath!);
    const reader = open(dbPath, extPath!);
    try {
      const workerQueue = worker.queue<EmailPayload>("emails", { visibilityTimeoutS: 45 });
      const readerQueue = reader.queue<EmailPayload>("emails", { visibilityTimeoutS: 45 });

      const id = workerQueue.enqueue({ recipient: "carol@example.com", template: "welcome" });
      const claimedBefore = now(worker);
      const job = workerQueue.claimOne("worker-7");
      const claimedAfter = now(worker);
      expect(job).not.toBeNull();

      const processing = readerQueue.getJob(id);
      expect(processing).not.toBeNull();
      expect(processing!.state).toBe("processing");
      expect(processing!.workerId).toBe("worker-7");
      expect(processing!.attempts).toBe(1);
      expect(processing!.claimExpiresAt).toBeGreaterThanOrEqual(claimedBefore + 45);
      expect(processing!.claimExpiresAt).toBeLessThanOrEqual(claimedAfter + 45);
      expect(processing!.payload).toEqual({
        recipient: "carol@example.com",
        template: "welcome",
      });

      expect(job!.ack()).toBe(true);
      expect(readerQueue.getJob(id)).toBeNull();
    } finally {
      reader.close();
      worker.close();
      rmSync(dir, { recursive: true, force: true });
    }
  });

  test(
    "a delayed job reports its runAt and is not claimable yet",
    withDb((db) => {
      const queue = db.queue<EmailPayload>("emails");

      const before = now(db);
      const id = queue.enqueue(
        { recipient: "dan@example.com", template: "receipt" },
        { delay: 60 },
      );
      const after = now(db);

      const snapshot = queue.getJob(id);
      expect(snapshot).not.toBeNull();
      expect(snapshot!.state).toBe("pending");
      expect(snapshot!.runAt).toBeGreaterThanOrEqual(before + 60);
      expect(snapshot!.runAt).toBeLessThanOrEqual(after + 60);
      expect(queue.claimOne("early-worker")).toBeNull();
    }),
  );

  test(
    "a retried job reports the new attempt count and run_at",
    withDb((db) => {
      const queue = db.queue<EmailPayload>("emails", { maxAttempts: 3 });
      const id = queue.enqueue({ recipient: "erin@example.com", template: "welcome" });

      const first = queue.claimOne("worker-1");
      expect(first!.attempts).toBe(1);
      expect(first!.maxAttempts).toBe(3);

      const retriedAt = now(db);
      expect(first!.retry(30, "boom")).toBe(true);

      const snapshot = queue.getJob(id);
      expect(snapshot!.state).toBe("pending");
      expect(snapshot!.attempts).toBe(1);
      expect(snapshot!.workerId).toBeNull();
      expect(snapshot!.runAt).toBeGreaterThanOrEqual(retriedAt + 30);
      expect(snapshot!.runAt).toBeLessThanOrEqual(now(db) + 30);
    }),
  );

  test(
    "the claim waker returns a job carrying the same detail",
    withDb(async (db) => {
      const queue = db.queue<EmailPayload>("emails", { visibilityTimeoutS: 60 });
      const id = queue.enqueue(
        { recipient: "frank@example.com", template: "receipt" },
        { priority: 3 },
      );

      const waker = queue.claimWaker({ idlePollS: 1 });
      try {
        const job = await waker.next("waker-worker");
        expect(job).not.toBeNull();
        expect(job!.id).toBe(id);
        expect(job!.state).toBe("processing");
        expect(job!.priority).toBe(3);
        expect(job!.workerId).toBe("waker-worker");
        expect(job!.payload.recipient).toBe("frank@example.com");
        expect(job!.ack()).toBe(true);
      } finally {
        waker.close();
      }
    }),
  );
});

File: packages/honker-go/typed_jobs_test.go
package honker

import (
	"path/filepath"
	"testing"
	"time"
)

// Issue #136: a claimed Job and a GetJob snapshot must both carry the
// full twelve-field job shape the core already returns, and a typed
// payload must survive the round trip through the database.

type emailPayload struct {
	Recipient string `json:"recipient"`
	Template  string `json:"template"`
	Version   int    `json:"version"`
}

// openTypedJobs returns two independent Database handles on one file:
// `worker` claims, `reader` reads. Separate handles mean separate
// connection pools, so the reader sees only what the worker committed.
func openTypedJobs(t *testing.T) (worker *Database, reader *Database) {
	t.Helper()
	extPath := findExtension(t)
	dbPath := filepath.Join(t.TempDir(), "typed.db")

	worker, err := Open(dbPath, extPath)
	if err != nil {
		t.Fatalf("open worker: %v", err)
	}
	t.Cleanup(func() { _ = worker.Close() })

	reader, err = Open(dbPath, extPath)
	if err != nil {
		t.Fatalf("open reader: %v", err)
	}
	t.Cleanup(func() { _ = reader.Close() })

	return worker, reader
}

func TestClaimedJobAndSnapshotCarryEveryField(t *testing.T) {
	worker, reader := openTypedJobs(t)
	wq := worker.Queue("emails", QueueOptions{MaxAttempts: 5, VisibilityTimeoutS: 120})
	rq := reader.Queue("emails", QueueOptions{MaxAttempts: 5, VisibilityTimeoutS: 120})

	expires := int64(600)
	before := time.Now().Unix()
	id, err := wq.Enqueue(
		emailPayload{Recipient: "alice@example.com", Template: "welcome", Version: 2},
		EnqueueOptions{Priority: 7, Expires: &expires},
	)
	if err != nil {
		t.Fatalf("enqueue: %v", err)
	}
	after := time.Now().Unix()
	if id <= 0 {
		t.Fatalf("enqueue returned id %d", id)
	}

	// ---- pending snapshot -------------------------------------------
	pending, err := rq.GetJob(id)
	if err != nil || pending == nil {
		t.Fatalf("GetJob(pending): row=%v err=%v", pending, err)
	}
	if pending.ID != id {
		t.Errorf("pending.ID = %d, want %d", pending.ID, id)
	}
	if pending.Queue != "emails" {
		t.Errorf("pending.Queue = %q, want %q", pending.Queue, "emails")
	}
	if pending.State != "pending" {
		t.Errorf("pending.State = %q, want %q", pending.State, "pending")
	}
	if pending.Priority != 7 {
		t.Errorf("pending.Priority = %d, want 7", pending.Priority)
	}
	if pending.RunAt < before || pending.RunAt > after {
		t.Errorf("pending.RunAt = %d, want within [%d, %d]", pending.RunAt, before, after)
	}
	if pending.WorkerID != nil {
		t.Errorf("pending.WorkerID = %q, want nil before any claim", *pending.WorkerID)
	}
	if pending.ClaimExpiresAt != nil {
		t.Errorf("pending.ClaimExpiresAt = %d, want nil before any claim", *pending.ClaimExpiresAt)
	}
	if pending.Attempts != 0 {
		t.Errorf("pending.Attempts = %d, want 0", pending.Attempts)
	}
	if pending.MaxAttempts != 5 {
		t.Errorf("pending.MaxAttempts = %d, want 5", pending.MaxAttempts)
	}
	if pending.CreatedAt < before || pending.CreatedAt > after {
		t.Errorf("pending.CreatedAt = %d, want within [%d, %d]", pending.CreatedAt, before, after)
	}
	if pending.ExpiresAt == nil {
		t.Fatalf("pending.ExpiresAt = nil, want run_at + %d", expires)
	}
	// enqueue derives run_at and expires_at from one unixepoch() read,
	// so the gap is exactly the requested TTL.
	if got := *pending.ExpiresAt - pending.RunAt; got != expires {
		t.Errorf("pending.ExpiresAt - RunAt = %d, want %d", got, expires)
	}
	pendingPayload, err := DecodePayload[emailPayload](pending.PayloadBytes())
	if err != nil {
		t.Fatalf("decode pending payload: %v", err)
	}
	want := emailPayload{Recipient: "alice@example.com", Template: "welcome", Version: 2}
	if pendingPayload != want {
		t.Errorf("pending payload = %+v, want %+v", pendingPayload, want)
	}

	// ---- claimed job -------------------------------------------------
	claimBefore := time.Now().Unix()
	job, err := wq.ClaimOne("worker-go")
	if err != nil || job == nil {
		t.Fatalf("ClaimOne: job=%v err=%v", job, err)
	}
	claimAfter := time.Now().Unix()

	if job.ID != id {
		t.Errorf("job.ID = %d, want %d", job.ID, id)
	}
	if job.Queue != "emails" {
		t.Errorf("job.Queue = %q, want %q", job.Queue, "emails")
	}
	if job.State != "processing" {
		t.Errorf("job.State = %q, want %q", job.State, "processing")
	}
	if job.Priority != 7 {
		t.Errorf("job.Priority = %d, want 7", job.Priority)
	}
	if job.RunAt != pending.RunAt {
		t.Errorf("job.RunAt = %d, want %d (unchanged by claim)", job.RunAt, pending.RunAt)
	}
	if job.WorkerID != "worker-go" {
		t.Errorf("job.WorkerID = %q, want %q", job.WorkerID, "worker-go")
	}
	if job.ClaimExpiresAt < claimBefore+120 || job.ClaimExpiresAt > claimAfter+120 {
		t.Errorf(
			"job.ClaimExpiresAt = %d, want within [%d, %d] (claim time + 120s timeout)",
			job.ClaimExpiresAt, claimBefore+120, claimAfter+120,
		)
	}
	if job.Attempts != 1 {
		t.Errorf("job.Attempts = %d, want 1 (claim increments)", job.Attempts)
	}
	if job.MaxAttempts != 5 {
		t.Errorf("job.MaxAttempts = %d, want 5", job.MaxAttempts)
	}
	if job.CreatedAt != pending.CreatedAt {
		t.Errorf("job.CreatedAt = %d, want %d (unchanged by claim)", job.CreatedAt, pending.CreatedAt)
	}
	if job.ExpiresAt == nil || *job.ExpiresAt != *pending.ExpiresAt {
		t.Errorf("job.ExpiresAt = %v, want %d", job.ExpiresAt, *pending.ExpiresAt)
	}
	claimedPayload, err := DecodePayload[emailPayload](job.Payload)
	if err != nil {
		t.Fatalf("decode claimed payload: %v", err)
	}
	if claimedPayload != want {
		t.Errorf("claimed payload = %+v, want %+v", claimedPayload, want)
	}

	// ---- the reader sees the processing details ----------------------
	processing, err := rq.GetJob(id)
	if err != nil || processing == nil {
		t.Fatalf("GetJob(processing): row=%v err=%v", processing, err)
	}
	if processing.State != "processing" {
		t.Errorf("processing.State = %q, want %q", processing.State, "processing")
	}
	if processing.WorkerID == nil || *processing.WorkerID != "worker-go" {
		t.Errorf("processing.WorkerID = %v, want %q", processing.WorkerID, "worker-go")
	}
	if processing.ClaimExpiresAt == nil || *processing.ClaimExpiresAt != job.ClaimExpiresAt {
		t.Errorf("processing.ClaimExpiresAt = %v, want %d", processing.ClaimExpiresAt, job.ClaimExpiresAt)
	}
	if processing.Attempts != 1 {
		t.Errorf("processing.Attempts = %d, want 1", processing.Attempts)
	}

	// ---- after ack the reader gets nothing ---------------------------
	acked, err := job.Ack()
	if err != nil || !acked {
		t.Fatalf("Ack: acked=%v err=%v", acked, err)
	}
	gone, err := rq.GetJob(id)
	if err != nil {
		t.Fatalf("GetJob after ack: %v", err)
	}
	if gone != nil {
		t.Fatalf("GetJob after ack = %+v, want nil", gone)
	}
}

func TestDelayedJobReportsItsRunAt(t *testing.T) {
	worker, _ := openTypedJobs(t)
	q := worker.Queue("delayed", QueueOptions{})

	delay := int64(3600)
	before := time.Now().Unix()
	id, err := q.Enqueue(map[string]any{"to": "later"}, EnqueueOptions{Delay: &delay})
	if err != nil {
		t.Fatalf("enqueue: %v", err)
	}
	after := time.Now().Unix()

	row, err := q.GetJob(id)
	if err != nil || row == nil {
		t.Fatalf("GetJob: row=%v err=%v", row, err)
	}
	if row.State != "pending" {
		t.Errorf("row.State = %q, want %q", row.State, "pending")
	}
	if row.RunAt < before+delay || row.RunAt > after+delay {
		t.Errorf("row.RunAt = %d, want within [%d, %d]", row.RunAt, before+delay, after+delay)
	}
	// Not claimable until run_at.
	job, err := q.ClaimOne("worker-go")
	if err != nil {
		t.Fatalf("ClaimOne: %v", err)
	}
	if job != nil {
		t.Fatalf("delayed job claimed early: %+v", job)
	}
}

// TestClaimedJobReportsBackDatedRunAt pins the claimed Job's RunAt and
// CreatedAt apart. An immediately-enqueued job has run_at == created_at
// to the second, so asserting both against the same instant proves
// nothing: the two fields can be transposed in the claim decode and the
// assertions still hold. Back-dating run_at with an absolute timestamp
// keeps the job claimable (a delayed job is not) while the two values
// differ by 100 seconds.
func TestClaimedJobReportsBackDatedRunAt(t *testing.T) {
	worker, reader := openTypedJobs(t)
	wq := worker.Queue("backdated", QueueOptions{MaxAttempts: 3, VisibilityTimeoutS: 60})
	rq := reader.Queue("backdated", QueueOptions{MaxAttempts: 3, VisibilityTimeoutS: 60})

	before := time.Now().Unix()
	past := before - 100
	id, err := wq.Enqueue(
		emailPayload{Recipient: "bob@example.com", Template: "reminder", Version: 1},
		EnqueueOptions{RunAt: &past},
	)
	if err != nil {
		t.Fatalf("enqueue: %v", err)
	}
	after := time.Now().Unix()

	// ---- snapshot keeps the two apart --------------------------------
	row, err := rq.GetJob(id)
	if err != nil || row == nil {
		t.Fatalf("GetJob: row=%v err=%v", row, err)
	}
	if row.RunAt != past {
		t.Errorf("row.RunAt = %d, want %d (the back-dated absolute run_at)", row.RunAt, past)
	}
	if row.CreatedAt < before || row.CreatedAt > after {
		t.Errorf("row.CreatedAt = %d, want within [%d, %d]", row.CreatedAt, before, after)
	}
	if row.RunAt == row.CreatedAt {
		t.Fatalf("row.RunAt == row.CreatedAt == %d, fixture no longer separates them", row.RunAt)
	}

	// ---- claimed job keeps the two apart -----------------------------
	job, err := wq.ClaimOne("worker-backdated")
	if err != nil {
		t.Fatalf("ClaimOne: %v", err)
	}
	if job == nil {
		t.Fatalf("ClaimOne returned nil, want the back-dated job to be claimable")
	}
	if job.RunAt != past {
		t.Errorf("job.RunAt = %d, want %d (the back-dated absolute run_at)", job.RunAt, past)
	}
	if job.CreatedAt < before || job.CreatedAt > after {
		t.Errorf("job.CreatedAt = %d, want within [%d, %d]", job.CreatedAt, before, after)
	}
	if job.RunAt == job.CreatedAt {
		t.Errorf("job.RunAt == job.CreatedAt == %d, want run_at %d and created_at ~%d",
			job.RunAt, past, before)
	}
	if job.CreatedAt-job.RunAt != row.CreatedAt-row.RunAt {
		t.Errorf(
			"claimed gap CreatedAt-RunAt = %d, snapshot gap = %d, want equal",
			job.CreatedAt-job.RunAt, row.CreatedAt-row.RunAt,
		)
	}
}

// TestDecodePayloadRejectsEmptyInput: an empty payload must not decode
// to a zero-valued T with a nil error. The core never emits an empty
// payload, so empty means the bytes never arrived, and a caller cannot
// otherwise tell that apart from a clean decode.
func TestDecodePayloadRejectsEmptyInput(t *testing.T) {
	for _, tc := range []struct {
		name    string
		payload []byte
	}{
		{"nil", nil},
		{"empty", []byte{}},
	} {
		t.Run(tc.name, func(t *testing.T) {
			got, err := DecodePayload[emailPayload](tc.payload)
			if err == nil {
				t.Fatalf("DecodePayload(%s) = %+v, nil — want an error", tc.name, got)
			}
			if got != (emailPayload{}) {
				t.Errorf("DecodePayload(%s) value = %+v, want the zero value", tc.name, got)
			}
		})
	}

	// A real payload still decodes.
	want := emailPayload{Recipient: "carol@example.com", Template: "digest", Version: 3}
	got, err := DecodePayload[emailPayload]([]byte(`{"recipient":"carol@example.com","template":"digest","version":3}`))
	if err != nil {
		t.Fatalf("DecodePayload(valid): %v", err)
	}
	if got != want {
		t.Errorf("DecodePayload(valid) = %+v, want %+v", got, want)
	}
}
